File: Week-02/Day-04/main.go
package main

import "fmt"

// Node represents a node in a circular linked list
type Node struct {
	data int
	next *Node
}

// CircularLinkedList represents a circular linked list
type CircularLinkedList struct {
	head *Node
}

func main() {
	list := &CircularLinkedList{}
	list.InsertAtBeginning(69)
	list.InsertAtEnd(7)
	list.InsertAtEnd(71)
	list.InsertAtBeginning(5)

	fmt.Println("Circular Linked List: ")
	list.PrintList()

	list.DeleteNode(5)
	fmt.Println("List after deleting node with data 5: ")
	list.PrintList()
}

// InsertAtBeginning inserts a new node at the beginning of the list
func (list *CircularLinkedList) InsertAtBeginning(data int) {
	newNode := &Node{data: data}
	if list.head == nil {
		list.head = newNode
		newNode.next = list.head
	} else {
		temp := list.head
		for temp.next != list.head {
			temp = temp.next
		}
		temp.next = newNode
		newNode.next = list.head
	}
}

// InsertAtEnd inserts a new node at the end of the list
func (list *CircularLinkedList) InsertAtEnd(data int) {
	newNode := &Node{data: data}
	if list.head == nil {
		list.head = newNode
		newNode.next = list.head
	} else {
		temp := list.head
		for temp.next != list.head {
			temp = temp.next
		}
		temp.next = newNode
		newNode.next = list.head
	}
}

// DeleteNode deletes a node with specific data from the list
func (list *CircularLinkedList) DeleteNode(data int) {
	if list.head == nil {
		fmt.Println("List is empty bro")
		return
	}

	temp := list.head
	var prev *Node
	for temp.data != data {
		if temp.next == list.head {
			fmt.Println("Node not found")
			return
		}
		prev = temp
		temp = temp.next
	}
	if temp == list.head {
		prev = prev.next
		for prev.next != list.head {
			prev = prev.next
		}
		list.head = temp.next
		prev.next = list.head
	} else if temp.next == list.head {
		prev.next = list.head
	} else {
		prev.next = temp.next
	}
}

// PrintList prints all the nodes in the list
func (list *CircularLinkedList) PrintList() {
	if list.head == nil {
		fmt.Println("List is empty.")
		return
	}
	temp := list.head
	for {
		fmt.Printf("%d -> ", temp.data)
		temp = temp.next
		if temp == list.head {
			break
		}
	}
	fmt.Println(temp.data) // Print the data of the last node pointing back to the head
}

File: Week-02/Day-04/main.cpp
#include <iostream>
using namespace std;

class DoublyLinkedList {
private:
  struct Node{
  int data;
  Node* next;
  Node* prev;
  Node(int val): data(val), next(nullptr), prev(nullptr) {}
  };
  Node* head;

public:
  DoublyLinkedList(): head(nullptr) {}
  ~DoublyLinkedList();

  void insertAtBeginning(int val);
  void insertAtEnd(int val);
  void insertAtPosition(int data, int position);
  void deleteNode(Node* node);
  void printList();
};

DoublyLinkedList::~DoublyLinkedList(){
  Node* current = head;
  Node* nextNode;
  while (current != nullptr) {
    nextNode = current->next;
    delete current;
    current = nextNode; 
  }
}

void DoublyLinkedList::insertAtBeginning(int val){
  Node* newNode = new Node(val);
  if (!head) {
    head = newNode;
  }
  if(head) {
    head->prev = newNode;
  }
    newNode->next = head;
    head = newNode;
}

void DoublyLinkedList::insertAtEnd(int val){
  Node* newNode = new Node(val);
  if(!head){
    head = newNode;
    return;
  }
  Node* temp = head;
  while(temp->next){
    temp = temp->next;
  }
  temp->next = newNode;
  newNode->prev = temp;
}

void DoublyLinkedList::insertAtPosition(int data, int position){
  Node* newNode = new Node(data);
  if (position == 0) {
    insertAtBeginning(data);
    return;
  } 
  Node* temp = head;
  for(int i = 0; temp && i < position-1; ++i){
    temp = temp->next;
  }
  if (!temp) return;
  newNode->next = temp->next;
  if(temp->next){
    temp->next->prev = newNode;
  }
  temp->next = newNode;
  newNode->prev = temp;
}

void DoublyLinkedList::deleteNode(Node* node){
  if(!head || !node) return;
  if(head == node){
    head = node->next;
  }
  if(node->next) {
    node->next->prev = node->next;
  }
  if (node->prev){
    node->prev->next = node->next;
  }
  delete node;
}

void DoublyLinkedList::printList(){
  Node* temp = head;
  while (temp != nullptr) {
    cout << temp->data << " ";
    temp = temp->next;
  }
  cout << endl;
}

int main(){

  DoublyLinkedList list;

  list.insertAtEnd(21);
  list.insertAtBeginning(71);
  list.insertAtPosition(2, 2);

//   list.deleteNode(head);
  list.printList();

}

